challenge-298/e-choroba/perl/ch-1.pl

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/perl
+use warnings;
+use strict;
+use feature qw{ say };
+use experimental qw( signatures );
+
+use List::Util qw{ min };
+
+use enum qw( HORIZONTAL VERTICAL SQUARE );
+
+=head1 THE ALGORITHM
+
+Start from the upper left corner. Store 3 numbers for each cell:
+
+1. How far we can go to the left without stepping on a 0 (HORIZONTAL);
+
+2. How far we can go up without stepping on a 0 (VERTICAL);
+
+3. How large square we can build leftwards and upwards (SQUARE).
+
+Getting HORIZONTAL and VERTICAL is easy: just add 1 to the neighbour's value
+in the respective direction. To get SQUARE, use 1 + SQUARE in the left-up
+diagonal neighbour, but if HORIZONTAL or VERTICAL for this cell is less, use
+it instead.
+
+ 1  1  1  1
+ 1  1  1  0
+ 1  1  1 (1)
+
+ [1 1 1][2 1 1][3 1  1 ][4    1  1]
+ [1 2 1][2 2 2][3 2 (2)][0    0  0]
+ [1 3 1][2 3 2][3 3  3 ][(4) (1) 1]
+                                 |
+                                 v
+                           min(2 + 1, 4, 1)]
+
+=cut
+
+sub maximal_square($matrix) {
+    my @strips;
+    my @starts;
+    my $maxsize = 0;
+    for my $y (0 .. $#$matrix) {
+        for my $x (0 .. $#{ $matrix->[0] }) {
+            if ($matrix->[$y][$x]) {
+                $strips[$y][$x][HORIZONTAL]
+                    = $x ? 1 + $strips[$y][ $x - 1 ][HORIZONTAL]
+                         : 1;
+                $strips[$y][$x][VERTICAL]
+                    = $y ? 1 + $strips[ $y - 1 ][$x][VERTICAL]
+                         : 0;
```

Comment:
@choroba : Is this a typo?

Reply:
@jo-37: Yes, thanks. I rewrote the algorithm several times (it originally started from the bottom right instead of top left), so it probably happened in one of the rewrites. Unfortunately, there was no test to catch it. See #11275.

Reply:
@choroba: Awesome approach!